MovieTime/MovieTime/Core/MTServiceHelper.swift

```diff
@@ -0,0 +1,57 @@
+//
+//  MTServiceHelper.swift
+//  MovieTime
+//
+//  Created by Enrique Melgarejo on 08/01/18.
+//  Copyright © 2018 Choynowski. All rights reserved.
+//
+
+import Foundation
+
+protocol MTServiceHelperProtocol {
+    func buildURL(for baseURL: URL?, pathComponent: String, and parameters: MTParameters?...) -> URL?
+}
+
+struct MTServiceHelper: MTServiceHelperProtocol {
+
+    func buildURL(for baseURL: URL?, pathComponent: String, and parameters: MTParameters?...) -> URL? {
+        guard let url = baseURL?.appendingPathComponent(pathComponent) else { return nil }
+
+        var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
+
+        if !parameters.isEmpty {
+            var queryItems: [URLQueryItem] = components?.queryItems ??  []
+            for params in parameters {
+                queryItems.append(contentsOf: buildQueryItems(parameters: params))
+            }
+            components?.queryItems = queryItems
+        }
+        return components?.url
+    }
+
+    private func buildQueryItems(parameters: MTParameters?) -> [URLQueryItem] {
+        guard let parameters = parameters else { return [] }
+
+        var queryItems: [URLQueryItem] = []
+        for (key, value) in parameters {
+
+            let queryName = String(describing: key)
+            var queryValue = ""
```

Comment:
in this context, how can it be avoided to declare a mutable property with an initial value before it is actually used?

Reply:
In this context it is possible to replace the `var queryValue = ""` for `var queryValue = String()`, since **queryValue** can be modified several times.
```
for (_, arrayValue) in array.enumerated() {
    let queryArrayValue = String(describing: arrayValue)
    queryItems.append(URLQueryItem(name: queryName, value: queryArrayValue))
}
```
If the string was assigned only once inside the if or else, it would be possible to have something like this:
```
let queryValue: String
if *someValidation* {
    queryValue = "Some value"
} else {
    queryValue = "Other value"
}
```

Reply:
I think it is indeed assigned only once in the scope it is declared so an immutable String could be used

Reply:
You're right. I changed the use of `queryValue` when the value is Array:
```
 +       if let array = value as? [Any] {
 +
 +                // For array objects
 +                for (_, arrayValue) in array.enumerated() {
 +                    let queryArrayValue = String(describing: arrayValue)
 +                    queryItems.append(URLQueryItem(name: queryName, value: queryArrayValue))
 +                }
 +            }
```

And now, I only use it inside `else if` validation. So this can be modified to:

```
    private func buildQueryItems(parameters: MTParameters?) -> [URLQueryItem] {
        guard let parameters = parameters else { return [] }

        var queryItems: [URLQueryItem] = []
        for (key, value) in parameters {

            let queryName = String(describing: key)

            if let array = value as? [Any] {

                // For array objects
                for (_, arrayValue) in array.enumerated() {
                    let queryArrayValue = String(describing: arrayValue)
                    queryItems.append(URLQueryItem(name: queryName, value: queryArrayValue))
                }
            } else if !(value is NSNull) {

                // For any objects
                let queryValue = String(describing: value)
                queryItems.append(URLQueryItem(name: queryName, value: queryValue))
            }
        }
        return queryItems
    }
````